lib/src/flutter_downloader.dart

```diff
@@ -0,0 +1,136 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:crypto/crypto.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_downloader/flutter_downloader.dart';
+import 'package:flutter_downloader/src/dart_download.dart';
+import 'package:flutter_downloader/src/platform_download.dart';
+
+part 'legacy_api.dart';
+
+/// The target where the download should be visible for the user.
+enum DownloadTarget {
+  /// Put the download in the download folder. If the target platform does
+  /// support download folders internal will be used.
+  downloadsFolder,
+
+  /// Put the download in the desktop folder. If the target platform does
+  /// support download folders internal will be used.
+  desktopFolder,
+
+  /// Put the download into the internal storage of the app.
+  internal,
+}
+
+/// Abstraction to observe download status updates
+abstract class DownloadProgress extends ChangeNotifier {
+  /// The state of the download
+  DownloadStatus get status;
+
+  /// The current progress in per mille [0...1000]
+  // If you need percent simply device it by ten. Per mille is used here for
+  // big downloads where you would see big jumps between progress steps.
+  int get progress;
+}
+
+/// Factory interface
+typedef CustomHttpClientFactory = HttpClient Function();
+
+/// Provides access to all functions of the plugin in a single place.
+/// The change notifier should just inform about changes of the
+/// [DownloadProgress].
+class FlutterDownloader extends ChangeNotifier
+    with FlutterDownloaderLegacy
+    implements DownloadProgress {
+  /// Get the flutter downloader.
+  factory FlutterDownloader() {
+    // Singleton is in Legacy part since it is required there too
+    return FlutterDownloaderLegacy._singleton;
+  }
+
+  FlutterDownloader._internal();
+
+  /// Add a custom http factory for platforms without platform specific
+  /// implementations like Android and iOS.
+  static CustomHttpClientFactory? customHttpClientFactory;
+
+  @override
+  int get progress => throw UnimplementedError();
+
+  @override
+  DownloadStatus get status => throw UnimplementedError();
+
+  // TODO check if that is required in the future
+  static bool _debug = false;
+
+  /// If true, more logs are printed.
+  static bool get debug => _debug;
+
+  /// Start a new download. The [Download] instance encapsulates the download
+  /// status and controls like cancel, pause and resume.
```

Comment:
I think it'd be nice to mention "it's recommended to pass a user agent if you're using a public, open API". No strong opinion though :)